feature: Single url, all networks

Matthews3301 opened the pull request:
# Description

Moving away from using subdomains for separate networks. More info here: https://www.notion.so/One-app-many-networks-0ec226994c1542c4af7a1d4be6be943b

1. Change network by using dropdown - it reloads the page, however, the url doesn't require subdomain anymore.
2. Open specific network by using url query string, e.g. `?networkId=137`.


## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Dependency changes
- [ ] Code refactor / cleanup
- [ ] Documentation or wording changes
- [ ] Other

## How should this be tested?

Change network using networks dropdown.

## Visual context

No changes.

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have requested at least 2 reviews (If the PR is significant enough, use best judgement here)
- [x] I have commented my code where relevant, particularly in hard-to-understand areas
- [n/a] If package-lock.json has changes, it was intentional.
- [x] The base of this PR is `master` if hotfix, `develop` if not

garethfuller commented:
> Open specific network by using url query string, e.g. ?networkId=137.
Could we not just use URL params? I've kind of suggested that in the URL structure table in notion. Also, lets stick to friendly network names instead of ids in the URLs please.

garethfuller commented:
On switching networks, it looks like tokenlists aren't being changed? I haven't looked at the code yet, but we'll also have to essentially reset the token registry which is controlled from tokens.provider.ts.
<img width="336" alt="Screenshot 2022-08-17 at 15 31 20" src="https://user-images.githubusercontent.com/2406506/185161385-482e5971-b379-41a5-ba11-ed403eba9633.png">

garethfuller commented:
When switching network on a pool page and redirecting back to the home page of the new network, please use the notification system like Pon suggested to say something like, Pool doesn't exist on <old-network>.

garethfuller commented:
When switching network on pages like /:network/trade, please don't redirect back to the home page. Instead please switch to the same page but with the new network. The only place we should need a redirect is on the pool pages.

Matthews3301 commented:
@garethfuller thank you, updated. The pool notifcation uses `useAlerts()` - do we use any other notification components? Or should I add one?

timjrobinson commented:
@garethfuller why do the trade page or homepage need the network included in the url? Pools do because there can be different pools of the same ID on different networks. Trade and homepage could use whatever is selected in the corner, or whatever network the users wallet is connect to. Uniswap doesn't include the network in the URL on the trade page.

garethfuller commented:
> @garethfuller thank you, updated. The pool notifcation uses `useAlerts()` - do we use any other notification components? Or should I add one?

@Matthews3301 `useNotifications()` is what Pon meant for you to use I think.

Matthews3301 commented:
@garethfuller Updated to `useNotifications()`

timjrobinson commented on src/plugins/router.ts:
```diff
@@ -59,103 +65,108 @@ const routes: RouteRecordRaw[] = [
-  },
-  {
-    path: '/vebal',
+    path: '/:networkSlug/vebal',
```
This page is only on mainnet, same with `get-vebal` and `unlock`

Matthews3301 replied:
Yes, I was thinking about this - I see what you mean. However, would it not be easier for users to understand this via the existing modal which explains that it's eth only? Might be confusing that all other urls start with the network name. 

Also, this is more future-proof if there is veBal on other networks.

garethfuller replied:
Yeah, I think it's fine to maintain the network in the URL for the verbal pages for the same points that Matthew made.

garethfuller commented:
@Matthews3301 https://www.loom.com/share/64a14377a1e246199e2ac7f25e5b69e4

Matthews3301 commented on src/composables/useNetwork.ts:
```diff
@@ -1,18 +1,30 @@
+    ? (Number(localStorage.getItem('networkId')) as Network)
+    : undefined;
+
+const NETWORK_ID =
```
Are you sure? This scenario happens in pretty much only case - when user opens `/` route without a network selected in url. Should we not show Ethereum by default to highlight our main pools like like 80-20 BAT-ETH and Aave boosted?

garethfuller replied:
imo I don't think we should be detecting the user's wallet's network and auto redirecting. I think that could be especially confusing if they follow a link to a pool on a specific network, for example: A user who's wallet is connected to mainnet clicks on an external URL linking to a polygon pool, on landing on the page we detect their network is mainnet and throw them back to the home page.

timjrobinson replied:
@garethfuller The users wallet would only come after the URL in priority. Priority should be URL flag -> local storage network -> users wallet currently selected network -> mainnet. 

@Matthews3301 A lot of people only use Polygon (or L2's) avoid mainnet because it's too expensive for them. It would make more sense for them to just be shown the Polygon homepage as their default landing page. 

It's also useful for repeat Balancer users - so that they don't have to continually select their favorite network every time they open Balancer. Though maybe this is covered by localStorageNetworkId.

garethfuller replied:
I still don't agree that we should auto-redirect based on a user's wallet (except in the case that we talked about Matthew, where a user is already connected to the correct network and then switches).

I do agree that we should store their last used network in localStorage and use that, but I'm not sure how. One issue for example, similar to what I described above: what if the user last used polygon and then left Balancer, they then click on a link to a mainnet pool page or to the arbitrum pools list from an external app, do we redirect them? We shouldn't. So in what cases would we use the localStorage variable to redirect on landing?

timjrobinson replied:
I think the localStorage network or using users wallet network only makes sense on the main pools list or trade page without any network in the URL. Which seems to be the logic here, having the network in the URL overrides all.

garethfuller commented on src/components/navs/AppNav/AppNavNetworkSelect.vue:
```diff
@@ -1,66 +1,106 @@
 }
 
 function appUrl(network: NetworkOption): string {
-  return `https://${network.subdomain}.balancer.fi`;
+  if (
+    router.currentRoute.value.name === 'pool' ||
+    router.currentRoute.value.name === 'create-pool' ||
+    router.currentRoute.value.name === 'invest' ||
+    router.currentRoute.value.name === 'withdraw' ||
+    router.currentRoute.value.name === 'migrate-pool'
+  )
+    return `/#/${network.networkSlug}?poolNetworkAlert=${networkConfig.name}`;
+
+  const currentRoute = router.currentRoute.value;
+  return router.resolve({
+    name: currentRoute.name ?? 'home',
+    params: { ...currentRoute.params, networkSlug: network.networkSlug },
+    query: currentRoute.query,
+  }).href;
 }
```
I imagine this function is actually no longer relevant? It was only used before to get the app domain for a given network, not the path.

Matthews3301 replied:
It performs a different role now, for pool routes it redirects to "home" and for all other routes it updates the current url with new network.

garethfuller replied:
Let's rename it then. Maybe something like `transformUrl` or `handleNetworkChange`?

garethfuller commented on src/composables/useNetwork.ts:
```diff
@@ -1,18 +1,30 @@
+
+const NETWORK_ID =
+  localStorageNetworkId ||
+  Number(process.env.VUE_APP_NETWORK) ||
```
VUE_APP_NETWORK should be deprecated with these changes. It shouldn't be required anymore.

Matthews3301 replied:
I think we still need it for unit tests... Not sure if possible to run some of the tests which require kovan without it.

garethfuller commented on src/composables/useNetwork.ts:
```diff
@@ -26,9 +38,6 @@ export const isTestnet = computed(() => isKovan.value || isGoerli.value);
 /**
  * METHODS
  */
-export function setNetworkId(id: Network) {
-  networkId.value = id;
-}
```
Should we not keep the possibilty to change the networkId programmatically? Ideally we have a single source of truth for what the current app network is, e.g. `networkId`. When a user lands on a route it's set to the relevant network, when a user switches network with the network switcher it's set to the relevant network. Then eventually we can transition everything to be reactive so we don't need refreshes.

Matthews3301 replied:
Do you mean should we keep this in preparation for future changes? Currently, there's no point for networkId to be reactive as network is always constant after an app is loaded.

garethfuller replied:
Yeah I guess that's what I was getting at, but actually, I think it's fine to remove it for now. We should focus on getting this working with a page reload first.

garethfuller commented on src/providers/token-lists.provider.ts:
```diff
@@ -34,6 +35,7 @@ export interface TokenListsProviderResponse {
 
 /** SETUP */
 const { uris } = tokenListService;
+const { NETWORK_ID } = useNetwork();
```
We should always be accessing a reactive property for the network. e.g. networkID.

garethfuller commented on src/providers/token-lists.provider.ts:
```diff
@@ -50,7 +52,7 @@ export default {
     let allTokenLists = {};
     try {
-      allTokenLists = require<TokenListMap>('/public/data/tokenlists.json');
+      allTokenLists = require<TokenListMap>(`/public/data/tokenlists/tokens-${NETWORK_ID}.json`);
```
This will be why some token icons are not displaying. `allTokenLists` now needs to be a reactive property that updates when `networkId.value` updates.

Matthews3301 replied:
The change in this line should fix the token icon problem. Does it still persist?

garethfuller reviewed (changes requested):
Good first pass.

garethfuller commented:
@Matthews3301 next week, please click around the app and perform actions whilst watching the console. Any errors that come up lets look into them and make sure they're not caused by network switching.

garethfuller commented:
Also, if the environment variable `VUE_APP_ENV` / `APP_ENV` when fetched from configService is either 'development' or 'staging' lets add Goerli to the network switcher.

garethfuller commented:
@Matthews3301 Load the app, then change network, then go to trade and try and perform a trade, it throws an error because of an incorrect network somewhere.

Matthews3301 commented:
@garethfuller By any chance, is this not this error (`underlying network changed`): https://docs.google.com/document/d/1l5cTqDaI7pt1QdNU7qL4j_cQtRD0HosFH6qa_suRabU/edit

This seems to be unrelated to this merge request but is a problem we should fix. I'll look into this more.

Matthews3301 commented:
@pkattera @garethfuller I added network change watcher so that if user has already connected their wallet and changed then changed their network through MetaMask, the page will be redirected to that new network. Ready for review.

Matthews3301 commented:
@garethfuller There is now a redirect which checks for a subdomain and redirects to the new format if necessary.

garethfuller commented on src/composables/useNavigationGuards.ts:
```diff
@@ -13,6 +17,53 @@ export default function useNavigationGuards() {
+    const networkFromSubdomain = networkFromSlug(subdomain);
+    if (networkFromSubdomain) {
+      // old format subdomain redirect
+      const networkFromUrl = networkFromSlug(to.params.networkSlug?.toString());
+      localStorage.setItem(
+        'networkId',
+        (networkFromUrl || networkFromSubdomain).toString()
+      );
+      window.location.href = networkFromUrl
+        ? `${appUrl()}${to.fullPath}`
+        : `${appUrl()}/${config[networkFromUrl || networkFromSubdomain].slug}${
+            to.fullPath
+          }`;
```
This should be a testable function that returns a URL. Please write some tests so we can be sure a variety of different input URLs will be redirected to the right place.

Matthews3301 replied:
👍 Refactored the code to split into separate functions and added unit tests. 

Separately, I don't think file `/src/lib/config/test.json` is used for anything - can I delete it?

garethfuller commented on src/composables/useNavigationGuards.ts:
```diff
@@ -13,6 +17,53 @@ export default function useNavigationGuards() {
+          }`;
+    } else {
+      // check for network in url and redirect if necessary
+      const networkSlug = to.params.networkSlug?.toString();
+      if (networkSlug) {
+        const networkFromUrl = networkFromSlug(networkSlug);
+        const localStorageNetwork: Network = networkFor(
+          localStorage.getItem('networkId') ?? '1'
+        );
+        if (!networkFromUrl) {
+          // missing or incorrect network name -> next() withtout network change
+          next();
+        } else if (localStorageNetwork === networkFromUrl) {
+          // if on the correct network -> next()
+          next();
+        } else {
+          // if on different network -> update localstorage and reload
+          document.write('');
+          localStorage.setItem('networkId', networkFromUrl.toString());
+          window.location.href = `/#${to.fullPath}`;
+          router.go(0);
+        }
```
Again, I think this should be extracted into a function and write tests to be sure it does as intended based on some test URLs.

Matthews3301 replied:
See above.

garethfuller commented on src/composables/useNavigationGuards.ts:
```diff
@@ -1,18 +1,98 @@
+  return betaEnv ? url.split('.')[1] : url.split('.')[0];
+}
+
+// old format subdomain redirect - e.g. "https://polygon.balancer.fi/"
+// if conflicting formats - use the url - e.g. "https://polygon.balancer.fi/#/arbitrum" -> arbitrum
+export function getOldFormatRedirectUrl(
+  networkFromSubdomain: Network,
+  networkSlug: string | undefined,
+  fullPath: string
+) {
+  const networkFromUrl = networkFromSlug(networkSlug ?? '');
+  localStorage.setItem(
+    'networkId',
+    (networkFromUrl || networkFromSubdomain).toString()
+  );
+  return networkFromUrl
+    ? `${appUrl()}${fullPath}`
+    : `${appUrl()}/${config[networkFromSubdomain].slug}${fullPath}`;
+}
```
Again, not super clear from the code what's going on here. I feel like the inputs here are over-complicating things. We should have a single function which takes the current URL and outputs the redirect URL if required

```suggestion
/**
 * Given current URL returns redirect URL, if no redirect required returns input URL.
 *
 * @param {string} url - Current URL.
 * @returns {string} URL to redirect too or current URL if no redirect required.
 */
export function getSubdomainNetworkRedirect(url: string): string {
  const subdomainNetwork = getSubdomainNetwork(url);
  if (subdomainNetwork) {
    // Construct redirect URL
  } else {
    return url;
  }
}
```
With a function like this we can more easily and clearly test; given input URL I expect this output URL.

Matthews3301 replied:
Thanks. How would we get fullPath and networkSlug from a url? By using regex?

I changed the function to this format but it requires more parameters.

garethfuller commented on src/composables/useNetwork.spec.ts:
```diff
@@ -0,0 +1,119 @@
+        result:
+          'https://localhost:8080/#/goerli/pool/0x5a6a8cffb4347ff7fc484bf5f0f8a2e234d34255000200000000000000000275',
+      },
+      {
+        url: 'https://ethereum.balancer.fi/#/portfolio',
+        result: 'https://localhost:8080/#/ethereum/portfolio',
+      },
```
The ethereum subdomain doesn't exist. Mainnet is on app.balancer.fi. I'd like to see a test case that insures `app.balancer.fi/#/some/path` basically doesn't redirect. Or redirects to `app.balancer.fi/#/ethereum/some/path` if it needs to.

I also noticed that `app.balancer.fi/#/trade`, for example, doesn't work. It doesn't redirect and it just loads the home page. That means any external link to out current mainnet trade page will be broken. We need to handle stuff like this.

Note: I've referenced `app.balancer.fi` as the domain here instead of the preview URL to keep it simple.

Matthews3301 commented on src/composables/useNetwork.ts:
```diff
@@ -47,38 +63,94 @@ export function networkFor(key: string | number): Network {
+  const subdomain = url.split('.')[0];
+  if (subdomain === 'beta') {
+    return url.split('.')[1];
+  } else {
```
Updated

garethfuller commented on src/composables/useNavigationGuards.ts:
```diff
@@ -1,18 +1,90 @@
+  /**
+   * Network url redirects - e.g. "app.balancer.fi/#/arbitrum".
+   */
+  router.beforeEach((to, from, next) => {
+    const urlNetwork = networkFromSlug(to.params.networkSlug?.toString() ?? '');
+    if (urlNetwork) {
+      const noNetworkChangeCallback = () => next();
+      const networkChangeCallback = (urlNetwork?: Network) => {
+        setNetworkAndRefresh(
+          (urlNetwork ?? '').toString(),
+          `/#${to.fullPath}`,
+          router
+        );
+      };
+      handleNetworkUrl(
+        to.params.networkSlug.toString(),
+        noNetworkChangeCallback,
+        networkChangeCallback
+      );
+    } else {
+      const nonNetworkedRoutes = [
+        '/',
+        '/terms-of-use',
+        '/privacy-policy',
+        '/cookies-policy',
+      ];
+      if (to.redirectedFrom || !nonNetworkedRoutes.includes(to.fullPath)) {
+        localStorage.setItem('networkId', networkId.value.toString());
+        router.push({
+          path: `/${networkSlug}${to.redirectedFrom?.fullPath ?? to.fullPath}`,
+        });
+      } else {
+        next();
+      }
+    }
+  });
```
@Matthews3301 Could you explain specifically what redirect cases this is handling? Just in the comments below. I'm trying to understand it but it's not super clear.

Matthews3301 replied:
1. E.g. `xxx.fi/polygon/trade` -> if current network != polygon, reload page with right network.
2. E.g. `xxx.fi/trade`(previous format) -> unless it's one of nonNetworkedRoutes, redirect to `xxx.fi/<current_network>/trade`.
